core/src/main/java/org/elasticsearch/action/search/InitialSearchPhase.java

```diff
@@ -131,55 +139,94 @@ private void onShardFailure(final int shardIndex, @Nullable ShardRouting shard,
 
     @Override
     public final void run() throws IOException {
-        boolean success = shardExecutionIndex.compareAndSet(0, maxConcurrentShardRequests);
-        assert success;
-        for (int i = 0; i < maxConcurrentShardRequests; i++) {
-            SearchShardIterator shardRoutings = shardsIts.get(i);
-            if (shardRoutings.skip()) {
-                skipShard(shardRoutings);
-            } else {
-                performPhaseOnShard(i, shardRoutings, shardRoutings.nextOrNull());
+        for (final SearchShardIterator iterator : skippedShardsIts) {
+            if (iterator.skip()) {
+                skipShard(iterator);
             }
         }
-    }
-
-    private void maybeExecuteNext() {
-        final int index = shardExecutionIndex.getAndIncrement();
-        if (index < shardsIts.size()) {
-            // we have to go fork execution to another thread or we could stack overflow on deeply recursive calls
-            executor.execute(() -> {
+        if (shardsIts.size() > 0) {
+            int maxConcurrentShardRequests = Math.min(this.maxConcurrentShardRequests, shardsIts.size());
+            final boolean success = shardExecutionIndex.compareAndSet(0, maxConcurrentShardRequests);
+            assert success;
+            for (int index = 0; index < maxConcurrentShardRequests; index++) {
                 final SearchShardIterator shardRoutings = shardsIts.get(index);
-                if (shardRoutings.skip()) {
-                    skipShard(shardRoutings);
-                } else {
+                if (!shardRoutings.skip()) {
```

Comment:
I also think we should make this an assertion

Reply:
Yes.